### PR TITLE
Introduce page-builder infrastructure with Elementor support

### PR DESCRIPTION
**Pull request by danieliser**
## Summary

Introduces the shared provider registry, capability interfaces, lifecycle coordinator, signed preview service, and isolated popup canvas, then migrates Elementor as the reference provider.

- Opens Popup Maker popups directly in Elementor.
- Uses the real themed popup canvas for editing and previewing.
- Preserves Elementor document rendering and interactive frontend widgets.
- Centralizes authorization, content routing, preview URLs, and secondary asset batching.
- Preserves the existing Bricks and Divi integrations unchanged; their provider migrations are separate child PRs.

This PR contains shared infrastructure plus Elementor only: 20 files changed. It does not add a Bricks, Divi, or other new builder provider.

## Verification

```text
composer run strauss: PASS
PHPCS on changed PHP files: 6 / 6, 0 issues
Aggregate PHPUnit with Bricks active: 937 tests, 2066 assertions, 2 skipped — PASS
Aggregate PHPUnit with Bricks absent: 937 tests, 2008 assertions, 14 skipped — PASS
PHPStan: 28 pre-existing errors, 0 in changed files
```

## Stack

Base: `develop`.
Next: Bricks in #1294.

### CODE REVIEW

**Comment by danieliser** [comment]
@codex review

File: assets/js/src/integration/bricks-preview.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+	if ( 'MutationObserver' in window ) {
+		new window.MutationObserver( adoptBuilderArea ).observe(
+			document.body,
+			{
+				childList: true,
+				subtree: true,
+			}
+		);
+	}
```
[the remark this answers is not in the document]
Comment (danieliser) [reply assets/js/src/integration/bricks-preview.js]: Not changing this without a reproduced editor regression. The observer converges through its existing insertion guards, and the real Bricks editor was exercised while editing and dragging content without a responsiveness problem. Adding a frame-delayed scheduling branch would be speculative behavior and conflicts with the verified-minimum pass.

File: assets/js/src/integration/elementor.js [hunk 2]
```diff
@@ -5,24 +5,49 @@
...
+				.find( '[data-element_type]' )
+				.addBack( '[data-element_type]' )
+				.each( function () {
+					window.elementorFrontend.elementsHandler.runReadyTrigger(
+						this
+					);
+				} );
 		}
-	);
+	};
```
[the remark this answers is not in the document]
Comment (danieliser) [reply assets/js/src/integration/elementor.js:30]: The one-time initialization marker was tested and removed during ablation because it preserved Elementor's broken zero-width carousel initialization. With the current visible-open trigger, the carousel measured correctly and advanced exactly one slide across repeated open cycles, so duplicate handlers were not reproduced. The repeated visible-state refresh is retained.

File: tests/php/tests/Bricks_Provider_Test.php [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+		$this->assertTrue( \Bricks\Helpers::is_post_type_supported( $popup_id ) );
+
+		// phpcs:disable WordPress.Security.NonceVerification.Recommended -- Test fixture.
+		$previous_get = $_GET;
+		$_GET         = [
+			'p'              => (string) $popup_id,
+			'bricks_preview' => (string) time(),
+		];
+		$builders     = \PopupMaker\plugin()->get_controller( 'Builders' );
+		$builders->providers()->register( $this->provider );
```
[the remark this answers is not in the document]
Comment (danieliser) [reply tests/php/tests/Bricks_Provider_Test.php]: No additional cleanup is needed. WordPress's WP_UnitTestCase calls _backup_hooks() during setup and _restore_hooks() during teardown, restoring wp_filter and the action/filter counters after every test. Adding a production reset method solely for this test would duplicate the framework lifecycle.

File: tests/php/tests/Builder_Providers_Test.php [hunk 4]
```diff
@@ -0,0 +1,957 @@
...
+			 */
+			protected function default_providers() {
+				return $this->stubs;
+			}
+		};
+
+		$controller->init();
+
+		return $controller;
+	}
```
[the remark this answers is not in the document]
Comment (danieliser) [reply tests/php/tests/Builder_Providers_Test.php:600]: No coordinator lifecycle method is needed for these tests. WordPress's WP_UnitTestCase snapshots all hooks in setup and restores them in teardown via _backup_hooks()/_restore_hooks(), so coordinators created by make_coordinator() do not leak their actions or filters into the next test.

File: classes/Builders/Divi.php [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+	 *
+	 * @return void
+	 */
+	public function register_hooks() {
+		add_filter( 'pum_popup_content', [ $this, 'render_visual_builder_mount' ], -12000, 2 );
+		add_filter( 'et_builder_post_types', [ $this, 'add_popup_support' ] );
+		add_filter( 'et_fb_is_enabled', [ $this, 'enable_frontend_builder' ], 10, 2 );
+
+		// Divi 4 disables the block editor for its builder post types.
+		add_filter( 'use_block_editor_for_post_type', [ $this, 'force_classic_editor' ], 999, 2 );
```
[the remark this answers is not in the document]
Comment (danieliser) [reply classes/Builders/Divi.php]: Confirmed and fixed in da9ab566. Runtime before the change on Divi 4.27.4 showed `built=yes type_filter=block post_filter=block`; the callback needed a post but was registered on the post-type filter. It is now registered on `use_block_editor_for_post`, validates the supplied `WP_Post` defensively, and scopes the classic-editor override to Divi-owned popup documents. Runtime after the change: `built=yes type_filter=block post_filter=classic`. Focused PHPUnit: `OK (1 test, 16 assertions)`.

File: classes/Builders/Elementor.php [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+		 * appear live. Elsewhere the display variant is correct, since it applies
+		 * the filters a visitor would see.
+		 */
+		if (
+			$popup_id === $this->get_canvas_popup_id() &&
+			method_exists( $frontend, 'get_builder_content' )
+		) {
+			$rendered = $frontend->get_builder_content( $popup_id );
+		} elseif ( method_exists( $frontend, 'get_builder_content_for_display' ) ) {
+			$rendered = $frontend->get_builder_content_for_display( $popup_id );
```
[the remark this answers is not in the document]
Comment (danieliser) [reply classes/Builders/Elementor.php:209]: Verified against the installed Elementor 4.2.1 source and live output; no change is warranted. `Frontend::get_builder_content()` calls `$css_file->enqueue()` whenever the document renders, independently of `$with_css`. The flag only controls the later `$css_file->print_css()` inline path. With the provider still passing the default `false`, the ordinary front page emits `<link id="elementor-post-33-css" ...>` in `<head>` and popup 33 is styled. Passing `true` would add inline document CSS to markup that already has its stylesheet enqueued, duplicating output.

File: templates/single-popup.php [hunk 7]
```diff
@@ -1,11 +1,188 @@
...
+	\PopupMaker\set_current_popup( $previous_popup );
+	wp_footer();
+	?>
+	<script id="pum-builder-preview-script">
+		( function ( $, PUM ) {
+			'use strict';
+
+			if ( ! $ || ! PUM || 'function' !== typeof PUM.getPopup ) {
+				return;
+			}
```
[the remark this answers is not in the document]
Comment (danieliser) [reply templates/single-popup.php]: Confirmed and fixed in ddcb7076. The existing guard could not run if the unqualified identifier itself threw; the IIFE now receives `window.jQuery`, so an absent dependency reaches the existing no-op branch. PHPCS passes for the template, and the required Popup Maker combined-script cache was cleared after rebasing the fix into the runnable stack.

File: classes/Controllers/Builders.php [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+	 * @param mixed $query_vars Parsed query variables.
+	 *
+	 * @return mixed Filtered query variables.
+	 */
+	public function allow_builder_request( $query_vars ) {
+		if ( ! is_array( $query_vars ) ) {
+			return $query_vars;
+		}
+
+		$popup_id = $this->get_edit_popup_id();
```
[the remark this answers is not in the document]
Comment (danieliser) [reply classes/Controllers/Builders.php:283]: Not reproduced and no change made. Popup Maker registers `popup` as non-publicly-queryable, so its runtime permalink is numeric: `get_permalink(15)` returned `http://localhost:8981/?post_type=popup&p=15`. The installed admin row emits Bricks as `/?post_type=popup&p=15&bricks=run`; the Divi row likewise emits `/?post_type=popup&p=65&et_fb_activation_nonce=...`. Those real entry paths provide the ID before the `request` filter, and both editors were mounted from those links. Adding slug resolution would support a request shape these integrations do not generate and would widen the provider contract without runtime need.

File: classes/Builders/Elementor.php [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+		/**
+		 * Inside the builder canvas the document must render unfiltered so edits
+		 * appear live. Elsewhere the display variant is correct, since it applies
+		 * the filters a visitor would see.
+		 */
+		if (
+			$popup_id === $this->get_canvas_popup_id() &&
+			method_exists( $frontend, 'get_builder_content' )
+		) {
+			$rendered = $frontend->get_builder_content( $popup_id );
```
[the remark this answers is not in the document]
Comment (danieliser) [reply classes/Builders/Elementor.php:207]: Verified against Elementor 4.2.1; the suggested change would blank the standalone preview, so no change was made. A signed popup preview is the primary queried document. Elementor guards `get_builder_content_for_display()` when `get_the_ID() === $post_id`. Runtime with popup 33 as the current post produced `current=33 raw_bytes=999 display_bytes=0 raw_has_heading=yes display_has_heading=no`. The retained canvas path is therefore required for both Elementor’s iframe and Popup Maker’s standalone primary-document preview. Elementor’s raw method still applies its `elementor/frontend/the_content` filter in this installed version.

**Comment by danieliser** [comment]
@codex review

**Comment by danieliser** [comment]
@codex review

File: classes/Builders/Elementor.php [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+/**
+ * Elementor support for popup documents.
+ *
+ * Elementor defaults to supporting posts and pages only, so this provider adds
+ * runtime support for popups. It also restores the popup query for Elementor's
+ * preview iframe and renders popup documents through Elementor's API.
+ *
+ * @since 1.25.0
+ */
+class Elementor implements BuilderProvider, EditsPopups, RendersDocuments, ProvidesPreviewUrl {
```
[the remark this answers is not in the document]
Comment (danieliser) [reply classes/Builders/Elementor.php:28]: Reproduced this exact lifecycle against Elementor 4.2.1 on a classic Twenty Seventeen page: the popup was excluded from preload, first discovered after the closing head tag, and Elementor enqueued elementor-post-315-css during late rendering. WordPress print_late_styles() emitted that stylesheet in the footer, and the real Chrome render visibly showed the popup container background, 44px heading, and theme styling. Elementor therefore does not need LoadsDocumentAssets for this path; adding it would declare a capability the runtime does not require. The shared canvas extraction was tested afterward in both Elementor and Bricks, with both editors visibly mounting and styled.

**Comment by danieliser** [comment]
Superseded by the simplified page-builder foundation merged in #1300 and the focused builder integrations that follow it.
